lib/CL/clEnqueueNDRangeKernel.c

```diff
@@ -129,81 +217,212 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
+          TRY_LEAST_GRAIN(z, x, y) else
+          TRY_LEAST_GRAIN(y, z, x) else
+          TRY_LEAST_GRAIN(x, y, z)
+
+          /* No luck either? Give up, kill everything */
+          DESPERATE_CASE(z) else DESPERATE_CASE(y) else DESPERATE_CASE(x)
+#undef DESPERATE_CASE
+#undef TRY_LEAST_GRAIN
+#undef TRY_HALVE
+        }
 
-        /* Finally, start setting them to zero starting from the Z
-           dimension. */
-        if (local_z > 1)
-          {
-            local_z = 1;
-            continue;
-          }
-        else if (local_y > 1)
-          {
-            local_y = 1;
-            continue;
-          }
-        else if (local_x > 1)
-          {
-            local_x = 1;
+      /* We now have the largest possible local work-group size that satisfies
+       * all the hard constraints (divide global, per-dimension bound, overall
+       * bound) and our soft constraint of being as close as possible a multiple
+       * of the preferred work-group size multiple. Such a greedy algorithm
+       * minimizes the total number of work-groups. In moderate-sized launch
+       * grid, this may result in less work-groups than the number of Compute
+       * Units, with a resulting imbalance in the workload distribution.
+       * At the same time, we want to avoid too many work-groups, since some
+       * devices are penalized by such fragmentation.
+       * Finding a good balance between the two is a hard problem, and generally
+       * depends on the device as well as the kernel utilization of its
+       * resources.
+       * Lacking that, as a first step we will simply try to guarantee that we
+       * have at least one work-group per CU, as long as the local work size
+       * does not drop below a given threshold.
+       */
+
+      /* Pick a minimum work-group size of 4 times the preferred work-group size
+       * multiple, under the assumption that this would be a good candidate
+       * below which a Compute Unit will not do enough work. */
+      const size_t min_group_size = 4*preferred_wg_multiple;
+
+      /* We need the number of Compute Units in the device, since we want
+       * at least that many work-groups, if possible */
+
+      cl_uint ncus; /* number of compute units */
+      prop_err = POname(clGetDeviceInfo)(command_queue->device,
+        CL_DEVICE_MAX_COMPUTE_UNITS, sizeof(cl_uint), &ncus, NULL);
+
+      if (prop_err != CL_SUCCESS) /* unlikely */
+        ncus = 1;
+
+      /* number of workgroups */
+      size_t nwg_x = global_x/local_x;
+      size_t nwg_y = global_y/local_y;
+      size_t nwg_z = global_z/local_z;
+
+      size_t splits; /* number of splits to bring ngws to reach ncu */
+      /* Only proceed if splitting wouldn't bring us below the minimum
+       * group size */
+      while (((splits = ncus/(nwg_x*nwg_y*nwg_z)) > 1) &&
+             (local_x*local_y*local_z > splits*min_group_size))
+        {
+          /* Very simple splitting approach: find a dimension divisible
```

Comment:
I think this is OK, but a tad better would be to favour splits that maintain as many dimensions as possible the preferred SIMD width -- but this can be optimized forever.